[C10D] Introduce C++ side Collective Callbacks.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109739
* __->__ #110307
* #111072
* #111061

C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

kumpera committed:
[C10D] Introduce C++ side Collective Callbacks.

C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -292,9 +292,40 @@ void _register_builtin_comm_hook(
+  write(sync_pipe, &m, 1);
+}
+
+void enable_event_collection(int pipe) {
```
I think this should be renamed to `enable_python_event_collection' or something that has python in the name.  I had to double check, but i believe python/c++ callbacks are truly independent as they should be.  (in other words, you don't have to call this enabler to make your c++ callbacks work, and the c++ callbacks are in no way sharing infrastructure with the python one.  So the `event_queue_enabled` and `event_queue` vars should probably also have python in their name.

kumpera committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/Hooks.cpp:
```diff
@@ -15,44 +15,22 @@ namespace c10d {
-    event_queue.push_back(std::move(evt));
-    char m = 'x';
-    write(sync_pipe, &m, 1);
+  std::unique_lock<std::mutex> lock(callbacks_lock);
+  for (auto& cb : callbacks_list) {
+    cb(evt);
```
I am not sure if I am following why do we want to call the callback when we insert event into the queue?

wconstab replied:
this PR refactors so that there is no queue by default.

i think it would be clearer to just rename `enqueue_c10d_event` to `call_c10d_event_callbacks`, so ill make that change.

A queue is a possible implementation detail of a callback.  The python system uses a queue.  If another customer implemented callbacks they may choose to use a queue but they control it.

The requirement is the callback(s) that are registered need to act quickly and safely, delegating their real work to some other thread or process.

wconstab replied:
yea, in this case 

A) the python callback system uses its own python thread to 'drain' the python event queue, and all the c++ callback does is push events onto the python event queue.  Most of this implementation is in the previous PR, but it gets refactored a bit in this PR to make the c++ callback mechanism more generic

B) any other c++ callback that some user registers is expected to use their own queue and background thread if they need to do long-running things.

wconstab committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Introduce C++ side Collective Callbacks."


C++ side callbacks allow for advance users to get
access to the collective firehose.

It's worth mentioning and discussing the dire environment that those
callbacks are invoked. From either main thread of watchdog thread and
with a PTD lock held.

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/Hooks.hpp:
```diff
@@ -22,8 +18,32 @@ struct TORCH_API EventInfo {
+ *
+ * Callbacks must not block or run for a long period of time.
+ * They are invoked from threads are part of PyTorch's critical distributed
+ * infrastrucute. The recomended pattern is for callbacks to enqueue the events
+ * on some queue and have a separate thread process those events. If the
+ * callback deadlocks, it will hang the whole process and stop PyTorch from
+ * detecting failures. Do not call into CUDA.
```
Well, so true what if python thread hangs at GIL...

wconstab replied:
this comment refers to the c++ callback. the c++ callback must not hang, or it hangs our watchdog and stops our timeout mechanism.  That's why its pretty much only allowed to put events onto a queue.

the thread that reads from the queue (whether thats a c++ thread or a python thread) can hang if it wants to- that wouldn't affect our watchdog, it just affects whatever system is built on top of that callback.

fduwjj commented on torch/csrc/distributed/c10d/Hooks.cpp:
```diff
@@ -15,44 +15,22 @@ namespace c10d {
-    char m = 'x';
-    write(sync_pipe, &m, 1);
+  std::unique_lock<std::mutex> lock(callbacks_lock);
+  for (auto& cb : callbacks_list) {
```
nit: can we just use the full name `callback`? the name `cb` could mean thousand things.. Not strong opinion but just my thought. But since the program is short, should be fine.

wconstab replied:
yep. ill do it in a follow up PR. i have a list of cleanups from the earlier PR too.

fduwjj reviewed (approved):
LGTM. thanks for keep polishing it.

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot revert

wconstab commented:
@pytorchbot revert -m "this sits on top of another PR https://github.com/pytorch/pytorch/pull/111072 that needs to be reverted due to internal release testing failure / multisect blame"

wconstab commented:
@pytorchbot revert -m "this sits on top of another PR https://github.com/pytorch/pytorch/pull/111072 that needs to be reverted due to internal release testing failure / multisect blame" -c nosignal
